core/include/traccc/edm/details/device_container.hpp

```diff
@@ -34,6 +34,15 @@ class device_container
     /// Inherit all of the base class's constructors
     using base_type::base_type;
 
+    /// Constructor from a "view object"
+    ///
+    /// Note that it may also be a "data" or "buffer" object. It just needs to
+    /// look like a "view object".
+    ///
+    template <template <typename, typename> class view_t>
+    TRACCC_HOST_DEVICE device_container(const view_t<header_t, item_t>& view)
+        : base_type(view.headers, view.items) {}
```

Comment:
This is undefined behaviour if it is called on the host with host-inaccessible memory.

Reply:
:confused: What?

Yes, it's also bad if you give null pointers to this constructor. But I really don't understand what's you're getting to here. This is the exact same kind of constructor that we have for all of the device types in vecmem. Are those constructors somehow fundamentally flawed as well because they make it possible to trigger undefined behaviour with them?

Note that CUDA is pretty good with making your code crash if you try to access GPU-only memory in host code. So I'm not even convinced that UB is the right term here...

Reply:
Well, this was the reason why these types were so conservatively equipped with const-reference constructors in the first place, because they can be pretty dangerous in heterogeneous code. What I'm trying to say is: shouldn't this be a `TRACCC_DEVICE` function, rather than `TRACCC_HOST_DEVICE`? In what case would you actually want to use this on the host? :thinking:

Reply:
For any number of testing purposes. :confused: In general I see very little value in making a function `TRACCC_DEVICE` only. Not in this project...

Note that by definition when we use this constructor in SYCL code, it will be compiled into "host code", whatever CUDA qualifiers we put on it. So only thinking about the CUDA qualifiers here is pretty flawed to start with.